### PR TITLE
Feature/updates in tfidf

### LINKED ISSUE #125: [FEAT]  improve the TF-IDF content model
## Feature Description
Replace the current TF-IDF + cosine similarity pipeline in `content_model.py` with dense sentence embeddings using `sentence-transformers` (`all-MiniLM-L6-v2`), with TF-IDF retained as a lightweight fallback for cold environments.

## Problem it Solves
TF-IDF is a bag-of-words model — it matches on exact token overlap and completely misses semantic similarity. For example, a query for "wireless headphones" will score near zero against an item described as "Bluetooth audio over-ear" even though they mean the same thing. This hurts recommendation quality especially for short or paraphrased product descriptions, which are common in the existing dataset.

## Proposed Solution
Swap the vectorisation step in `content_model.py` to use `sentence-transformers`:

1. Add `sentence-transformers` to `requirements.txt`
2. In `content_model.py`, encode all item text fields (title + description) using `SentenceTransformer('all-MiniLM-L6-v2')`
3. Store the resulting dense embedding matrix (replace the sparse TF-IDF matrix)
4. Compute cosine similarity using `sklearn.metrics.pairwise.cosine_similarity` — the scoring step stays identical
5. Keep TF-IDF as a fallback flag (`USE_EMBEDDINGS=True` in `.env`) so the change is opt-in and doesn't break existing deployments
6. Add a benchmark in `evaluation.py` comparing Precision@K and NDCG@K between TF-IDF and embedding-based content scores

`all-MiniLM-L6-v2` is ~80MB, runs on CPU in under 50ms per batch, and requires no GPU — keeping the sub-50ms search SLA intact.

## Alternatives Considered
- **BM25 (`rank_bm25`)**: Better than plain TF-IDF for keyword relevance, still lexical. Good intermediate step but doesn't solve the semantic gap.
- **Fine-tuned BERT**: Too heavy for CPU inference at this scale; overkill for product title similarity.

## Additional Context
- Target file: `content_model.py`
- Related files: `requirements.txt`, `evaluation.py`, `.env.example`
- Model card: https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2
- The hybrid score formula (`α · content_score + β · collab_score + γ · sentiment_score`) stays unchanged — only the source of `content_score` changes, so `hybrid_model.py` needs no edits

### PR DESCRIPTION
**Pull request by parinaB**
## What changed

- Improved the TF-IDF recommendation pipeline in `content_model.py`
- Added stronger text preprocessing using normalized lowercase title mapping. 
- Optimized similarity scoring flow to avoid unnecessary computation issues
- Added duplicate filtering in recommendation results
- Improved recommendation quality by refining cosine similarity handling
- Added safer handling for missing/null text values in the combined column
- Fixed title lookup consistency issues caused by case-sensitive matching
- Refined preprocessing and similarity computation logic
- Improved search result formatting and metadata handling
- Added better debugging/testing support for local recommendation checks

## Why
The earlier TF-IDF implementation had multiple stability and recommendation-quality issues. These changes improve:
- recommendation relevance,
- system stability,
- dataset compatibility,
- and overall usability of the content-based recommender.

## How to test

```bash
pip install -r requirements.txt
python
```
Inside python 
```
import pandas as pd
from content_model import ContentRecommender

df = pd.read_csv("datasets/books.csv", low_memory=False)

df["title"] = df["Book-Title"]

df["combined"] = (
    df["Book-Title"].astype(str) + " " +
    df["Book-Author"].astype(str)
)

model = ContentRecommender(df)

print(model.recommend(df["title"].iloc[0], top_n=5))
```

Expected behavior
- No crashes/errors
- No duplicate recommendations
- Returned books should be contextually similar

## Screenshots (Proof)
Screenshot 2026-05-20 at 12 14 34 AM

< I did add the code to cancel self-recommendation in content_model, but didnt respond > 

## Checklist
- [✔] I have read the [CONTRIBUTING.md](CONTRIBUTING.md)
- [ ✔] My code follows PEP8 style (`flake8 .`)
- [✔ ] I have tested my changes locally
- [ ] I have added/updated tests where applicable
- [✔ ] I can explain every line of code I've written
- [✔ ] I have NOT used AI-generated code without understanding and attributing it

## Related issue
Closes #125 

## AI assistance disclosure

- [ ] I did not use AI assistance for this PR
- [✔] I used AI assistance for: Debugging Git workflow issues, Reviewing TF-IDF optimization approaches, Improving documentation and PR descriptions.

### CODE REVIEW

**Comment by parinaB** [comment]
@leonagoel could u kindly review when free...

**Review by leonagoel — APPROVED** [review]

**Comment by leonagoel** [comment]
✅ Marked as approved but **cannot auto-merge** — this PR has conflicts.

Please resolve conflicts and push again, then re-add the `status:approved` label.

**Review by leonagoel — CHANGES_REQUESTED** [review]
Thanks for the PR, @parinaB! The TF-IDF updates look interesting. However there are **conflicts in 7 files** and GitHub says they are **too complex to resolve in the web editor** — meaning you must resolve them via command line.

**Conflicting files:**
- `collaborative_model.py`
- `content_model.py`
- `data_adapter.py`
- `data_preprocessing.py`
- `dataset_manager.py`
- `evaluation.py`
- `hybrid_model.py`

Your branch is very far behind `main`. The easiest fix is to start fresh:
```
# Option 1 — Fresh branch (recommended)
git checkout main
git pull origin main
git checkout -b feature/updates-in-tfidf-fresh
# manually re-apply only your TF-IDF changes
git add .
git commit -m "feat: updates in tfidf"
git push origin feature/updates-in-tfidf-fresh
# then open a new PR from this branch
```

```
# Option 2 — Rebase (harder)
git fetch origin
git rebase origin/main
# resolve all 7 conflicts manually
git push --force-with-lease
```

Please address within **48 hours** or this PR will be closed.

**Comment by parinaB** [comment]
@leonagoel conflicts resolved. thank you so much for letting me know..

**Comment by parinaB** [comment]
@leonagoel pls review and if possible merge before conflicts arise if possible

**Comment by parinaB** [comment]
@leonagoel can u pls merge or provide changes if needed???

**Comment by parinaB** [comment]
any updatess????
